New step-functions-workflows-collection - bedrock-evaluations Initial commit

rohanmeh opened the pull request:
*Issue #, if available:*

*Description of changes:*
New workflow for orchestrating multiple Bedrock evaluations.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bfreiberg commented on bedrock-evaluations/README.md:
````diff
@@ -0,0 +1,69 @@
+    ```
+1. From the command line, use AWS SAM to deploy the AWS resources for the workflow as specified in the template.yaml file:
+    ```
+    sam deploy --guided
````
I needed to add --capabilities CAPABILITY_NAMED_IAM to the command to get it to deploy

bfreiberg replied:
Deployment failed with:

> Resource handler returned message: "Invalid request provided: CreateCollectionRequest(Description=Collection for vector search, Name=vector-search-collection, Tags=[Tag(Key=aws:cloudformation:stack-name, Value=bedrock-eval), Tag(Key=aws:cloudformation:logical-id, Value=OpenSearchCollection), Tag(Key=aws:cloudformation:stack-id, Value=arn:aws:cloudformation:eu-central-1:914766352016:stack/bedrock-eval/9f3638b0-e9e8-11ef-a14d-0a8c9768cfb5)], Type=VECTORSEARCH), No matching security policy of encryption type found for collection name: vector-search-collection. Please create security policy of encryption type for this collection. (Service: OpenSearchServerless, Status Code: 400, Request ID: 6ddb9760-e4e9-4e69-9e17-92e0e43ac569)" (RequestToken: 0519a234-9d90-a4be-458a-085ad0ead078, HandlerErrorCode: InvalidRequest)

bfreiberg replied:
That error is still occurring on a deploy

rohanmeh commented on bedrock-evaluations/template.yaml:
```diff
@@ -0,0 +1,649 @@
+                Action:
+                  - bedrock:InvokeModel
+                Resource:
+                  - 'arn:aws:bedrock:us-east-1::foundation-model/amazon.titan-embed-text-v2:0'
```
fixed in latest commit

rohanmeh commented on bedrock-evaluations/template.yaml:
```diff
@@ -0,0 +1,649 @@
+                Action:
+                  - bedrock:Retrieve
+                  - bedrock:RetrieveAndGenerate
+                Resource: !Sub 'arn:aws:bedrock:us-east-1:${AWS::AccountId}:knowledge-base/*'
```
fixed in latest commit

rohanmeh commented on bedrock-evaluations/template.yaml:
```diff
@@ -0,0 +1,649 @@
+            End: true
+      DefinitionSubstitutions:
+        IngestionEmbeddingModelArn: >-
+          arn:aws:bedrock:us-east-1::foundation-model/amazon.titan-embed-text-v2:0
```
fixed in latest commit

rohanmeh commented on bedrock-evaluations/README.md:
```diff
@@ -0,0 +1,69 @@
+This workflow creates multiple Knowledge Bases, performs an ingestion job into a vector database, and runs evaluation jobs for you to assess the effectiveness of your ingestion jobs. This is done by making use of Step Functions Parallel state to run multiple testing flows in parallel."
+
+## Image
+Provide an exported .png of the workflow in the `/resources` directory from [Workflow stuio](https://docs.aws.amazon.com/step-functions/latest/dg/workflow-studio.html) and add here.
```
fixed in latest commit

bfreiberg commented on bedrock-evaluations/README.md:
```diff
@@ -0,0 +1,68 @@
+
+This Amazon Step Functions workflow ingests data into multiple Knowledge Bases and evaluates the performance of your Knowledge Bases with your test suite of prompts.
+
+Learn more about this workflow at Step Functions workflows collection: << Add the live URL here >>
```
```suggestion
Learn more about this workflow at Step Functions workflows collection: https://serverlessland.com/workflows/bedrock-evaluations-sam
```

bfreiberg commented on bedrock-evaluations/README.md:
````diff
@@ -0,0 +1,68 @@
+    ```
+1. Change directory to the pattern directory:
+    ```
+    cd step-functions-workflows-collection/bedrock-evaluations
````
```suggestion
    cd step-functions-workflows-collection/bedrock-evaluations-sam
```

bfreiberg reviewed (changes requested):
I'm still facing the following error:

> Deployment failed with:
> 
> Resource handler returned message: "Invalid request provided: CreateCollectionRequest(Description=Collection for vector search, Name=vector-search-collection, Tags=[Tag(Key=aws:cloudformation:stack-name, Value=bedrock-eval), Tag(Key=aws:cloudformation:logical-id, Value=OpenSearchCollection), Tag(Key=aws:cloudformation:stack-id, Value=arn:aws:cloudformation:eu-central-1:914766352016:stack/bedrock-eval/9f3638b0-e9e8-11ef-a14d-0a8c9768cfb5)], Type=VECTORSEARCH), No matching security policy of encryption type found for collection name: vector-search-collection. Please create security policy of encryption type for this collection. (Service: OpenSearchServerless, Status Code: 400, Request ID: 6ddb9760-e4e9-4e69-9e17-92e0e43ac569)" (RequestToken: 0519a234-9d90-a4be-458a-085ad0ead078, HandlerErrorCode: InvalidRequest)

bfreiberg commented:
Please also note that there a few comments still unresolved in `bedrock-evaluations.json`

rohanmeh commented:
> I'm still facing the following error:
> 
> > Deployment failed with:
> > Resource handler returned message: "Invalid request provided: CreateCollectionRequest(Description=Collection for vector search, Name=vector-search-collection, Tags=[Tag(Key=aws:cloudformation:stack-name, Value=bedrock-eval), Tag(Key=aws:cloudformation:logical-id, Value=OpenSearchCollection), Tag(Key=aws:cloudformation:stack-id, Value=arn:aws:cloudformation:eu-central-1:914766352016:stack/bedrock-eval/9f3638b0-e9e8-11ef-a14d-0a8c9768cfb5)], Type=VECTORSEARCH), No matching security policy of encryption type found for collection name: vector-search-collection. Please create security policy of encryption type for this collection. (Service: OpenSearchServerless, Status Code: 400, Request ID: 6ddb9760-e4e9-4e69-9e17-92e0e43ac569)" (RequestToken: 0519a234-9d90-a4be-458a-085ad0ead078, HandlerErrorCode: InvalidRequest)

Apologies. I added a DependsOn line to the Opensearch Collection for the security policies to be created prior to the collection.

rohanmeh commented on bedrock-evaluations/statemachine/statemachine.asl.json:
```diff
@@ -0,0 +1,425 @@
+                "OutputDataConfig": {
+                  "S3Uri": "s3://knowledge-base-evaluations-evaluationbucket-lucehibx2avn/output/"
+                },
+                "RoleArn": "arn:aws:iam::637423582422:role/bedrock-eval-role"
```
outdated and updated in latest commit

rohanmeh commented:
Json file: Had previously updated the bucket names but forgot to resolve comments. Removed specific region mentions.

Updated readme with suggested changes.

bfreiberg commented:
Looks good, thank you for your contribution. Your workflow will be added soon to Serverlessland.com
